# 4.7: Deriving New Variables

Content List:
1. Import Libraries, Create Path and Import Datasets
2. Create Product Price Labels
3. Create Busy Day Labels
4. Create New Busy Day Labels
5. Create Busy Period of Day Labels
6. Export Updated Dataset

1. Import Libraries, Create Path and Import Datasets

In [1]:
#Importing Libraries
import pandas as pd
import numpy as nb
import os

In [2]:
#Create Path
path = r'C:\Users\ryanc\OneDrive\Documents\Data Analytics Exercises\Python\Instacart Basket Analysis'

In [3]:
#Import order products merged df
ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

2. Create Product Price Labels

In [5]:
#Create High Range Product Label
ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [6]:
#Create Mid Range Range Product Label
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [7]:
#Create Low Range Product Label
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [8]:
#View value counts of new labels
ords_prods_merged['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

3. Create Busy Day labels

In [9]:
#View frequency of order days of the week
ords_prods_merged['order_dow'].value_counts(dropna = False)

order_dow
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [11]:
#Assign labels for days of the week based on how busy they are
result = []

for value in ords_prods_merged["order_dow"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [12]:
#create new column 'busiest_day' for these new labels
ords_prods_merged['busiest_day'] = result

In [13]:
#View frequency of busiest_day labels
ords_prods_merged['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

4. Create New Busy Day Labels

In [18]:
##Question 2
#Change busiest days labels to include the 2 busiest days and the 2 slowest days
new_result = []

for value in ords_prods_merged["order_dow"]:
  if value == 0 or value == 1:
    new_result.append("Busiest days")
  elif value == 4 or value == 3:
    new_result.append("Slowest Days")
  else:
    new_result.append("Regularly Busy days")

In [19]:
#Create new column for new busiest days labels
ords_prods_merged['new_busiest_days'] = new_result

In [20]:
#View frequency of new_busiest_days labels
ords_prods_merged['new_busiest_days'].value_counts(dropna = False)

new_busiest_days
Regularly Busy days    12916111
Busiest days           11864412
Slowest Days            7624336
Name: count, dtype: int64

#Question 3: As we already know the values for how busy each day of the week is we can easily check our values to ensure they add up. The 2 slowest days are days 3 and 4 and correctly add up to 7,624,336 for our new "slowest_days" label. Days 0 and 1 are the busiest days and correctly add up 11,864,412. While the remaining days of Day 2, 5 and 6 add up to our "regularly_busy_days label at 12,916,111. This simple math ensures we have our values correct. 

In [21]:
#View our data
ords_prods_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   3139998      138    prior            28          6                 11   
1   1977647      138    prior            30          6                 17   
2    389851      709    prior             2          0                 21   
3    652770      764    prior             1          3                 13   
4   1813452      764    prior             3          4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day     new_busiest_days  
0  Mid-range product  Regularly busy  Regularly Busy days  
1  Mid-range product  Regularly busy  Regularly Busy days  
2  Mid-range product     Busiest day         Busiest days  
3  Mid-range product  Regularly busy         Slowest Days  
4  Mid-range product      Least busy         Slowest Days

5. Create Busy Period of Day Labels

In [22]:
#View freqency of orders by hour
ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [27]:
#View statistics of freqency of orders by hour
ords_prods_merged['order_hour_of_day'].value_counts(dropna = False).describe()

count    2.400000e+01
mean     1.350202e+06
std      1.078369e+06
min      5.128100e+04
25%      2.725620e+05
50%      1.117230e+06
75%      2.556034e+06
max      2.761760e+06
Name: count, dtype: float64

Determining busy and least busy time periods:
Viewing the descriptive statistics of the frequency of orders by hours shows us that that the average number of orders by hour is 1,350,202 orders. The standard deviation is 1,078,369 orders. Using these numbers we can determine that orders over 1 Std Dev away from the average are orders at or above 2,428,571. Orders 1 Std Dev below the average are orders at or below 271,833. Using this we see the hours between 9-16 are the busiest hours of the day, although not necessarily in order of 9 to 16 respectfully. These hours of the day all exceed 2,428,571 orders. The least busy hours are from 0-5, although again not necessarily in that order. This leaves the hours of between 6-8 and 17-23 as average orders. 


In [44]:
#Create labels using the time frames determined
hour_label = []

for value in ords_prods_merged["order_hour_of_day"]:
  if value in [9, 10, 11, 12, 13, 14, 15, 16]:
    hour_label.append("Most Orders")
  elif value in [0, 1, 2, 3, 4, 5]:
    hour_label.append("Fewest Orders")
  else:
    hour_label.append("Average Orders")

In [45]:
#Create new column for busiest hours labels
ords_prods_merged['busiest_hours'] = hour_label

In [46]:
ords_prods_merged['busiest_hours'].value_counts(dropna = False)

busiest_hours
Most Orders       21118071
Average Orders    10690460
Fewest Orders       596328
Name: count, dtype: int64

In [56]:
#View head of new column sorted by busiest_hours
ords_prods_merged[['order_hour_of_day', 'busiest_hours']].head(25).sort_values(by=['busiest_hours'])

order_hour_of_day   busiest_hours
1                  17  Average Orders
2                  21  Average Orders
4                  17  Average Orders
5                   7  Average Orders
21                 23  Average Orders
8                  20  Average Orders
18                 21  Average Orders
16                  8  Average Orders
15                  8  Average Orders
22                  3   Fewest Orders
20                  2   Fewest Orders
19                  0   Fewest Orders
17                  0   Fewest Orders
0                  11     Most Orders
12                 13     Most Orders
13                 12     Most Orders
23                 14     Most Orders
11                 13     Most Orders
10                 11     Most Orders
9                  12     Most Orders
7                  10     Most Orders
6                  14     Most Orders
3                  13     Most Orders
14                 10     Most Orders
24                 14     Most Orders

6. Export Updated Dataset

In [57]:
#Export Data
ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))